Enforce 500 character limit on mod rpc data

Prismism0 opened the pull request:
Chunk mod list and mod rpc list to maximum length of 500 characters to conform to network limits. This fixes the following error:
```
<<Script Error>>scripts/managers/mod/mod_manager.lua:615: Failed to pack parameter 3, too many characters in string with max length 500<</Script Error>>
<<Lua Stack>>  [1] =[C]: in function rpc_mod_user_data
```

This is becoming increasingly important as some mods have many rpcs and it only takes a handful of mods to exceed this limit.

I wrote this fix in such a way that all scenarios which previously work will still work, even if Steam has not updated VMF for one of the players. So long as the mod/rpc strings are less than 500 characters, the data sent over the network will be exactly the same as before. Only if the data is over 500 characters will the new logic take effect.

I was able to test this with one other person, but not more.

Vermintide-Analytics committed:
Enforce 500 character limit on mod rpc data

Vermintide-Analytics committed:
Refactor for performance

Also clear data in remove_player to handle case where we may never receive full pong data

Prismism0 commented:
I have not yet tested my latest changes, only verified that it builds and tested the string logic in an online lua tool. So I would not say this is ready yet, but I am fairly confident the changes can be reviewed anyway. Hopefully I can test soon.

ManuelBlanc commented on vmf/scripts/mods/vmf/modules/core/network.lua:
```diff
@@ -159,7 +161,21 @@ end
+		local mod_data_block = _shared_mods_map:sub(block_start, block_start + MAX_MOD_DATA_LENGTH - 1)
+		local rpc_data_block = _shared_rpcs_map:sub(block_start, block_start + MAX_MOD_DATA_LENGTH - 1)
+		
+		rpc_chat_message(peer_id, 4, Network.peer_id(), mod_data_block, rpc_data_block, false, true, false)
```
Consider getting `Network.peer_id()` just once

Prismism0 replied:
Done

ManuelBlanc commented on vmf/scripts/mods/vmf/modules/core/network.lua:
```diff
@@ -246,32 +262,20 @@ end
-      user_rpcs_dictionary[1] = cjson.decode(rpc_data1) -- mods
-      user_rpcs_dictionary[2] = cjson.decode(rpc_data2) -- rpcs
+      user_rpcs_dictionary[1] = cjson.decode(mod_data) -- mods
+      user_rpcs_dictionary[2] = cjson.decode(rpc_data) -- rpcs
```
Nice! I like the descriptive variable names. Maybe the comments can be deleted now since they don't add anything

Prismism0 replied:
Done

ManuelBlanc commented on vmf/scripts/mods/vmf/modules/core/network.lua:
```diff
@@ -289,7 +293,48 @@ local function vmf_network_recv(sender, channel_id, rpc_data1, rpc_data2)
 
+local expected_pong_data_blocks = {}
+local partial_pong_mod_data = {}
+local partial_pong_rpc_data = {}
```
The convention for file upvalues in VMF is to have them at the top of the file and prefix them with `_` like `_partial_pong_rpc_data`, etc.

Prismism0 replied:
Done

ManuelBlanc commented on vmf/scripts/mods/vmf/modules/core/network.lua:
```diff
@@ -289,7 +293,48 @@ local function vmf_network_recv(sender, channel_id, rpc_data1, rpc_data2)
+		if rpc_data1 then
+			partial_pong_mod_data[sender] = partial_pong_mod_data[sender] .. rpc_data1
+		end
+		if rpc_data2 then
```
These will never be nil: `string.sub` returns an empty string if the interval is empty.

Prismism0 replied:
Removed the `if`s. I also looked at http://www.lua.org/source/5.1/lvm.c.html#luaV_concat to verify that Lua optimizes the concatenation when the 2nd parameter is empty. Interestingly, it does not do the same if the 1st parameter is empty, but it shouldn't be a big deal.

ManuelBlanc reviewed (changes requested):
Looks good! Added some small comments that would be nice to address.
Thanks again for picking up this!

Vermintide-Analytics committed:
Address PR suggestions

1. Optimize Network.peer_id() calls
2. Remove redundant comments
3. Follow VMF style for file-level variables
4. Remove if statements where condition will always be true anyway

Vermintide-Analytics committed:
Remove accidental whitespace change

Dreomite commented on vmf/scripts/mods/vmf/modules/core/network.lua:
```diff
@@ -159,7 +165,22 @@ end
+  local total_blocks = math.ceil(math.max(#_shared_mods_map / MAX_MOD_DATA_LENGTH, #_shared_rpcs_map / MAX_MOD_DATA_LENGTH))
+  
+  if total_blocks > 1 then
+	rpc_chat_message(peer_id, 4, self_peer_id, "blocks", total_blocks, false, true, false)
```
The VMF project convention is to use spaces for indentation (2). Otherwise, the formatting looks skewed in many editors. For example, here's how it looks in mine:

![image](https://user-images.githubusercontent.com/35058624/236172715-8ffb0fb6-d3e6-498a-abf1-294e4c3305b7.png)  

It would be great if you changed all the Tabs to Spaces here and in the rest of the proposed changes.
 
If you're using VSCode, you can simply convert Tabs to Spaces by pressing `Ctrl+Shift+P` and selecting "Indent Using Spaces".

---

Also, if you're using VSCode, consider adding `"files.trimTrailingWhitespace": true` to your local config, so it automatically removes trailing whitespaces on file save. Because there are a few of those. This suggestion is optional.

Vermintide-Analytics replied:
Fixed

Dreomite reviewed (changes requested):
Thank you for dealing with this long-lingering issue!

Please, consider adding yourself to https://github.com/Vermintide-Mod-Framework/Vermintide-Mod-Framework/blob/master/CONTRIBUTORS.MD (In this or any other of your PRs)

Visually, everything looks good and ready for merging, except 1 thing that needs addressing.

Also, regarding testing. Unfortunately, due to circumstances, I'm not able to test the changes properly, so, after merging this PR, I'll update [VMF Beta](https://steamcommunity.com/sharedfiles/filedetails/?id=1500112422) workshop item. I hope you will be able to test it properly then, and, once it's done, I'll send the new version to Fatshark for sanctioning.

Vermintide-Analytics committed:
Whitespace standardization

Also add myself to contributors list
